common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/checksum/ShuffleChecksumHelper.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  private static long readChecksumByReduceId(File checksumFile, int reduceId) throws IOException {
+    try (DataInputStream in = new DataInputStream(new FileInputStream(checksumFile))) {
+      ByteStreams.skipFully(in, reduceId * 8);
+      return in.readLong();
+    }
+  }
+
+  private static long calculateChecksumForPartition(
+      ManagedBuffer partitionData,
+      Checksum checksumAlgo) throws IOException {
+    InputStream in = partitionData.createInputStream();
+    byte[] buffer = new byte[CHECKSUM_CALCULATION_BUFFER];
+    try(CheckedInputStream checksumIn = new CheckedInputStream(in, checksumAlgo)) {
+      while (checksumIn.read(buffer, 0, CHECKSUM_CALCULATION_BUFFER) != -1) {}
+      return checksumAlgo.getValue();
+    }
+  }
+
+  /**
+   * Diagnose the possible cause of the shuffle data corruption by verifying the shuffle checksums.
+   *
+   * There're 3 different kinds of checksums for the same shuffle partition:
+   *   - checksum (c1) that is calculated by the shuffle data reader
+   *   - checksum (c2) that is calculated by the shuffle data writer and stored in the checksum file
+   *   - checksum (c3) that is recalculated during diagnosis
+   *
+   * And the diagnosis mechanism works like this:
+   * If c2 != c3, we suspect the corruption is caused by the DISK_ISSUE. Otherwise, if c1 != c3,
+   * we suspect the corruption is caused by the NETWORK_ISSUE. Otherwise, the cause remains
+   * CHECKSUM_VERIFY_PASS. In case of the any other failures, the cause remains UNKNOWN_ISSUE.
+   *
+   * @param checksumFile The checksum file that written by the shuffle writer
+   * @param reduceId The reduceId of the shuffle block
+   * @param partitionData The partition data of the shuffle block
+   * @param checksumByReader The checksum value that calculated by the shuffle data reader
+   * @return The cause of data corruption
+   */
+  public static Cause diagnoseCorruption(
+      File checksumFile,
+      int reduceId,
+      ManagedBuffer partitionData,
+      long checksumByReader) {
+    Cause cause;
+    try {
+      long diagnoseStart = System.currentTimeMillis();
+      // Try to get the checksum instance before reading the checksum file so that
+      // `UnsupportedOperationException` can be thrown first before `FileNotFoundException`
+      // when the checksum algorithm isn't supported.
+      Checksum checksumAlgo = getChecksumByFileExtension(checksumFile.getName());
+      long checksumByWriter = readChecksumByReduceId(checksumFile, reduceId);
+      long checksumByReCalculation = calculateChecksumForPartition(partitionData, checksumAlgo);
+      long duration = System.currentTimeMillis() - diagnoseStart;
+      logger.info("Shuffle corruption diagnosis took {} ms, checksum file {}",
+        duration, checksumFile.getAbsolutePath());
+      if (checksumByWriter != checksumByReCalculation) {
+        cause = Cause.DISK_ISSUE;
+      } else if (checksumByWriter != checksumByReader) {
```

Comment:
I probably mentioned this before - can we move this to before we recompute checksum via `calculateChecksumForPartition` ?
It is a cheap check, and if network is the culprit, we immediately return.

Reply:
Yeah, I also replied somewhere. Let me repeat here.

I think it wouldn't work. Because there could be two reasons if we find `checksumByWriter != checksumByReader` at the beginning. As you know, `checksumByReader` is calculated base on the corrupted block at fetcher side. And the corruption can be caused by either network issue or disk issue. So we can't know the exact cause if we check this at the beginning.

Reply:
You are right, my mistake.